sdk/storage/azure-storage-blob/azure/storage/blob/_download.py

```diff
@@ -531,81 +529,112 @@ def chunks(self):
-        is unspecified or is -1, all bytes will be read.
+        Read the specified bytes or chars from the stream. If `encoding`
+        was specified on `download_blob`, it is recommended to use the
+        chars parameter to read a specific number of chars to avoid decoding
+        errors. If size/chars is unspecified or negative all bytes will be read.
 
-        :param Optional[int] size:
+        :param int size:
             The number of bytes to download from the stream. Leave unspecified
-            or set to -1 to download all bytes.
+            or set negative to download all bytes.
+        :param Optional[int] chars:
+            The number of chars to download from the stream. Leave unspecified
+            or set negative to download all chars. Note, this can only be used
+            when encoding is specified on `download_blob`.
         :returns:
             The requested data as bytes or a string if encoding was specified. If
             the return value is empty, there is no more data to read.
         :rtype: T
         """
-        if size == -1:
-            return self.readall()
         # Empty blob or already read to the end
         if size == 0 or self._offset >= self.size:
             return b'' if not self._encoding else ''
 
-        stream = BytesIO()
-        remaining_size = size
-
-        # Start by reading from current_content if there is data left
-        if self._offset < len(self._current_content):
-            start = self._offset
-            length = min(remaining_size, len(self._current_content) - self._offset)
-            read = stream.write(self._current_content[start:start + length])
-
-            remaining_size -= read
-            self._offset += read
-            if self._progress_hook:
-                self._progress_hook(self._offset, self.size)
-
-        if remaining_size > 0:
-            start_range = self._get_downloader_start_with_offset()
-
-            # End is the min between the remaining size, the file size, and the end of the specified range
-            end_range = min(start_range + remaining_size, self._file_size)
-            if self._end_range is not None:
-                end_range = min(end_range, self._end_range + 1)
+        if size > -1 and self._encoding:
+            warnings.warn(
+                "Size parameter specified with text encoding enabled. It is recommended to use chars "
+                "to read a specific number of characters instead."
+            )
+        if size > -1 and chars is not None:
+            raise ValueError("Cannot specify both size and chars.")
+        if not self._encoding and chars is not None:
+            raise ValueError("Must specify encoding to read chars.")
+        if self._text_mode and size > -1:
+            raise ValueError("Stream has been partially read in text mode. Please use chars.")
+
+        download_complete = self._download_offset >= self.size
+        if not self._text_mode and chars is not None:
```

Comment:
Yes, currently the first scenario would work because it would fall back into the old behavior where it reads in bytes but then decodes the result before returning. This would result in a `UnicodeError` if those 5 bytes could not be properly decoded. It would be a bit of an odd scenario to read some chars as bytes then switch to reading chars, but I didn't see a reason to block it if we could support it.

However, I think I realized this won't really work due to how we keep the offset internally. Internally the offset is in bytes OR chars depending on what you're reading so reading some bytes and then some chars would mean the offset would be incorrect. I am going to switch to making these completely mutually exclusive and block reading in chars mode once you've partially read in bytes mode.

Reply:
Yep sounds like a good plan.